### ISSUE TITLE
vkDestroyInstance leak

### ISSUE THREAD

**Issue by DJTECKING** [KhronosGroup/Vulkan-LoaderAndValidationLayers#1298]
Seen with valgrind that vkDestroyInstance doesn't frees the hole VkInstance structure.
Compiled with option -g and -lvulkan (only core).
Launched with valgrind options : --leak-check=full --show-leak-kinds=all
Seen :
```
==5928== Syscall param sendmsg(msg.msg_iov[0]) points to uninitialised byte(s)
==5928==    at 0x5160170: __sendmsg_nocancel (in /usr/lib64/libc-2.23.so)
==5928==    by 0x6008606: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==5928==    by 0x6004A68: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==5928==    by 0x5F9CB5D: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==5928==    by 0x74B6F55: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==5928==    by 0x74AF6D3: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==5928==    by 0x74B1332: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==5928==    by 0x74B1B33: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==5928==    by 0x5FF1869: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==5928==    by 0x74ADF83: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==5928==    by 0x74AE908: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==5928==    by 0x5F9CE05: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==5928==  Address 0xffeffeffc is on thread 1's stack
```

Witch is a nvidia issue (with no connection to vulkan I guess)

and
```
==5928== HEAP SUMMARY:
==5928==     in use at exit: 32 bytes in 1 blocks
==5928==   total heap usage: 2,807 allocs, 2,806 frees, 58,559,808 bytes allocated
==5928== 
==5928== 32 bytes in 1 blocks are still reachable in loss record 1 of 1
==5928==    at 0x4C2DA60: calloc (vg_replace_malloc.c:711)
==5928==    by 0x541F66E: _dlerror_run (in /usr/lib64/libdl-2.23.so)
==5928==    by 0x541EFA1: dlopen@@GLIBC_2.2.5 (in /usr/lib64/libdl-2.23.so)
==5928==    by 0x4E42471: ??? (in /usr/lib64/libvulkan.so.1.0.30)
==5928==    by 0x4E461DA: vkCreateInstance (in /usr/lib64/libvulkan.so.1.0.30)
==5928==    by 0x4008C5: main (vulkan.c:56)
==5928== 
==5928== LEAK SUMMARY:
==5928==    definitely lost: 0 bytes in 0 blocks
==5928==    indirectly lost: 0 bytes in 0 blocks
==5928==      possibly lost: 0 bytes in 0 blocks
==5928==    still reachable: 32 bytes in 1 blocks
==5928==         suppressed: 0 bytes in 0 blocks
```

Which looks like a memory leak.

Using Fedora24 On acer veriton and gainward nvidia GT710.

Thanks a lot.

**Comment by mark-lunarg** [comment]
This is seen when running without validation layers, correct?

**Comment by DJTECKING** [comment]
Yes, to be more specific, I use only old C, gcc -Wall -g -lvulkan, no ide, and no validation layers.

**Comment by MarkY-LunarG** [comment]
What program are you running?  I've run Cube and VIA on Fedora Core 25 (using the Mesa driver) and I don't see this.  I see other issues, but they appear to be in the Mesa driver for me.  Can you also try building Cube out of the "marky_cube_alloc" branch?  I've modified cube to use external allocator/free functions, and it tends to show me just the real leaks with the loader and drivers.

**Comment by DJTECKING** [comment]
I ran that code:
```
#include<vulkan/vulkan.h>
#include<stdlib.h>
#include<stdio.h>

int main(int argc, char *argv[]) {
	// Application's informations :
	VkApplicationInfo app_info = {0};
	// Informations on instance that we want to create :
	VkInstanceCreateInfo instance_info = {0};
	// Vulkan instance :
	VkInstance instance = {0};

	// Extentions that vulkan must use to interact with the window manager
	const char *const enabled_extension_names[2] = { // Seen in glfw/src/x11_window.c line 2418
		"VK_KHR_surface",
		"VK_KHR_xlib_surface"
	};

	// Application's informations filling :
	app_info.sType = VK_STRUCTURE_TYPE_APPLICATION_INFO;
	app_info.pNext = 0;
	app_info.pApplicationName = "vulkan test";
	app_info.applicationVersion = VK_MAKE_VERSION(0, 0, 0);
	app_info.pEngineName = "None";
	app_info.engineVersion = VK_MAKE_VERSION(0, 0, 0);
	app_info.apiVersion = VK_API_VERSION_1_0;

	// Instance informations filling :
	instance_info.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
	instance_info.pNext = 0;
	instance_info.flags = 0;
	instance_info.pApplicationInfo = &app_info;
	instance_info.enabledLayerCount = 0;
	instance_info.ppEnabledLayerNames = 0;
	// Informations on layers dans extentions here : https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/loader/LoaderAndLayerInterface.md
	instance_info.enabledExtensionCount = 2; // Seen in glfw/src/vulkan.c line 243
	instance_info.ppEnabledExtensionNames = enabled_extension_names;

	// Vulkan instance creation
	if(vkCreateInstance(&instance_info, 0, &instance) != VK_SUCCESS) {
		printf("Failed to create vulkan instance\n");
	    return -1;
	}

	// Vulkan instance destruction
	vkDestroyInstance(instance, 0);

    return 0;
}
```

I might have wrote errors too.

Compilation command :
`> gcc -Wall -g -o bin/vulkan_issue_1298 demo/vulkan_issue_1298.c -lvulkan`

Test command :
```
> valgrind --leak-check=full --show-leak-kinds=all ./bin/vulkan_issue_1298 
==2378== Memcheck, a memory error detector
==2378== Copyright (C) 2002-2015, and GNU GPL'd, by Julian Seward et al.
==2378== Using Valgrind-3.11.0 and LibVEX; rerun with -h for copyright info
==2378== Command: ./bin/vulkan_issue_1298
==2378== 
==2378== Syscall param sendmsg(msg.msg_iov[0]) points to uninitialised byte(s)
==2378==    at 0x5160170: __sendmsg_nocancel (in /usr/lib64/libc-2.23.so)
==2378==    by 0x6008606: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==2378==    by 0x6004A68: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==2378==    by 0x5F9CB5D: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==2378==    by 0x74B6F55: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==2378==    by 0x74AF6D3: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==2378==    by 0x74B1332: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==2378==    by 0x74B1B33: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==2378==    by 0x5FF1869: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==2378==    by 0x74ADF83: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==2378==    by 0x74AE908: ??? (in /usr/lib64/libnvidia-glcore.so.375.20)
==2378==    by 0x5F9CE05: ??? (in /usr/lib64/libGLX_nvidia.so.375.20)
==2378==  Address 0xffefff4ac is on thread 1's stack
==2378== 
==2378== 
==2378== HEAP SUMMARY:
==2378==     in use at exit: 32 bytes in 1 blocks
==2378==   total heap usage: 2,799 allocs, 2,798 frees, 58,559,844 bytes allocated
==2378== 
==2378== 32 bytes in 1 blocks are still reachable in loss record 1 of 1
==2378==    at 0x4C2DA60: calloc (vg_replace_malloc.c:711)
==2378==    by 0x541F66E: _dlerror_run (in /usr/lib64/libdl-2.23.so)
==2378==    by 0x541EFA1: dlopen@@GLIBC_2.2.5 (in /usr/lib64/libdl-2.23.so)
==2378==    by 0x4E42471: ??? (in /usr/lib64/libvulkan.so.1.0.30)
==2378==    by 0x4E461DA: vkCreateInstance (in /usr/lib64/libvulkan.so.1.0.30)
==2378==    by 0x40081B: main (vulkan_issue_1298.c:40)
==2378== 
==2378== LEAK SUMMARY:
==2378==    definitely lost: 0 bytes in 0 blocks
==2378==    indirectly lost: 0 bytes in 0 blocks
==2378==      possibly lost: 0 bytes in 0 blocks
==2378==    still reachable: 32 bytes in 1 blocks
==2378==         suppressed: 0 bytes in 0 blocks
==2378== 
==2378== For counts of detected and suppressed errors, rerun with: -v
==2378== Use --track-origins=yes to see where uninitialised values come from
==2378== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

I'm fetching your branch and compile it, I can also provide more details with the numbered allocation, it's size, it's content, and send it as a bin file and track it with gdb if you want. I just need to know before spending any time on investigation, if it will be helpful for you.

**Comment by karl-lunarg** [comment]
I was able to reproduce both problems on Ubuntu 16.04 with nvidia graphics.

Yes, the first problem about uninitialized bytes would be an nvidia driver issue, so there is nothing we can do here.

For the second problem, I verified that the string passed to dlopen() was allocated and stored in a structure exactly once and the exact same address was freed.  So the program logic looks fine.

There have been bugs in this area when calling dlopen() when using the pthread library.  See https://bugs.launchpad.net/ubuntu/+source/valgrind/+bug/1160352.  This bug report is over 3 years old and is marked as fixed.  But it seems to have cropped up again.  A more recent [report](https://github.com/baldurk/renderdoc/issues/286#issuecomment-227598483) suggests that at least a related problem still exists.

Conclusion is that we can't do much with these particular problems.  Closing.
